### PR TITLE
CPLAT-8122 Shorten propTypes boilerplate

### PR DESCRIPTION
**Pull request by aaronlademann-wf**
## Motivation
The `Component2.propTypes` boilerplate was unnecessarily verbose.

## Changes
1. Add a class that contains the metadata that was previously provided as separate arguments, and change the `PropValidator` typedef signature to only require two arguments - with the second argument being an instance of the metadata class.

    __Before:__

    ```dart
    @override
    get propTypes => {
      'name': (Map props, propName, componentName, location, propFullName) {
        if (props[propName].length > 20) {
          return ArgumentError('...');
        }
        return null;
      },
    };
    ```

    __After:__

    ```dart
    @override
    get propTypes => {
      'name': (Map props, info) {
        if (props[info.propName].length > 20) {
          return ArgumentError('...');
        }
        return null;
      },
    };
    ```

1. Update JS interop to pass through the field values from the `ReactValidatorInfo` fields as individual arguments to the JS `React.propTypes` function.
1. Update some dynamic call sites where they could be tightened to at least `Function`.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: @greglittlefield-wf @joebingham-wk @kealjones-wk @sydneyjodon-wk 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        - [ ] Passing CI
        - [ ] Passing CI on [the analogous over_react PR](https://github.com/Workiva/over_react/pull/414)

### CODE REVIEW

File: example/test/react_test_components.dart [hunk 1]
```diff
@@ -7,9 +7,10 @@ import "package:react/react_dom.dart" as react_dom;
...
   get propTypes => {
-        'name': (Map props, propName, componentName, location, propFullName) {
-          if (props[propName].length > 20) {
-            return ArgumentError('(${props[propName]}) is too long. $propName has a max length of 20 characters.');
+        'name': (Map props, info) {
+          String propValue = props[info.propName];
+          if (propValue.length > 20) {
+            return ArgumentError('($propValue) is too long. $propValue has a max length of 20 characters.');
           }
           return null;
```
Comment (greglittlefield-wf) [inline example/test/react_test_components.dart]: `return null` is no longer part of the boilerplate
```suggestion
```

File: lib/react_client/bridge.dart [hunk 2]
```diff
@@ -116,13 +119,16 @@ class Component2BridgeImpl extends Component2Bridge {
...
         String secret,
       ) {
         // Create a Dart consumable version of the JsMap.
         var convertedProps = JsBackedMap.fromJs(props);
         // Call the users validator with the newly wrapped props.
-        var error = validator(convertedProps, propName, componentName, location, propFullName);
+        var error = validator(
+            convertedProps,
+            PropValidatorInfo(
+                propName: propName, componentName: componentName, location: location, propFullName: propFullName));
```
Comment (greglittlefield-wf) [inline lib/react_client/bridge.dart]: #nit trailing commas might be nice here for formatting purposes
```suggestion
                propName: propName, componentName: componentName, location: location, propFullName: propFullName,));
```

File: lib/src/prop_validator.dart [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/// Typedef for `Component2.propTypes` - used to check the validity of one or more of the [props].
+///
+/// [info] is a [PropValidatorInfo] class that contains metadata about the prop referenced as
+/// the key within the `Component2.propTypes` map.
+/// `propName`, `componentName`, `location` and `propFullName` are available.
+typedef PropValidator<T> = Error Function(T props, PropValidatorInfo info);
```
Comment (greglittlefield-wf) [inline lib/src/prop_validator.dart]: This should be `dynamic` as per the AC, allowing `return null` to be omitted without there being any hints
```suggestion
typedef PropValidator<T> = dynamic Function(T props, PropValidatorInfo info);
```

File: test/lifecycle_test/component2.dart [hunk 4]
```diff
@@ -151,10 +151,11 @@ ReactDartComponentFactoryProxy2 PropTypesTest = react.registerComponent(() => ne
...
-        'intProp': (Map props, propName, componentName, location, propFullName) {
-          if (props[propName] is! int) {
+        'intProp': (Map props, info) {
+          dynamic propValue = props[info.propName];
+          if (propValue is! int) {
             return ArgumentError(
-                '$propName should be int. {"props": "$props", "propName": "$propName", "componentName": "$componentName", "location": "$location", "propFullName": "$propFullName"}');
+                '${info.propName} should be int. {"props": "$props", "propName": "${info.propName}", "componentName": "${info.componentName}", "location": "${info.location}", "propFullName": "${info.propFullName}"}');
           }
           return null;
```
Comment (greglittlefield-wf) [inline test/lifecycle_test/component2.dart]: `return null` is no longer part of the boilerplate
```suggestion
```

File: test/lifecycle_test/component2.dart [hunk 5]
```diff
@@ -151,10 +151,11 @@ ReactDartComponentFactoryProxy2 PropTypesTest = react.registerComponent(() => ne
 
 class PropTypesTestComponent extends react.Component2 {
   get propTypes => {
-        'intProp': (Map props, propName, componentName, location, propFullName) {
-          if (props[propName] is! int) {
+        'intProp': (Map props, info) {
+          dynamic propValue = props[info.propName];
```
Comment (greglittlefield-wf) [inline test/lifecycle_test/component2.dart]: #nit prefer using var for local variables
```suggestion
          var propValue = props[info.propName];
```

File: lib/src/prop_validator.dart [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+/// Metadata about a prop being validated by a [PropValidator].
+class PropValidatorInfo {
+  final String propName;
```
Comment (greglittlefield-wf) [inline lib/src/prop_validator.dart]: #nit might be nice to have doc comments for these, but not necessary

File: test/lifecycle_test/component2.dart (hunk 4 above)
Comment (aaronlademann-wf) [reply test/lifecycle_test/component2.dart]: nope

File: lib/src/prop_validator.dart (hunk 3 above)
Comment (aaronlademann-wf) [reply lib/src/prop_validator.dart]: nope

File: example/test/react_test_components.dart (hunk 1 above)
Comment (aaronlademann-wf) [reply example/test/react_test_components.dart]: nope

Comment (greglittlefield-wf) [reply example/test/react_test_components.dart]: 😿

File: lib/src/prop_validator.dart (hunk 3 above)
Comment (greglittlefield-wf) [reply lib/src/prop_validator.dart]: 😿

Comment (kealjones-wk) [reply lib/src/prop_validator.dart]: its so weird, it like like implicitly types it based on the conditional return... whattt.

**Comment by kealjones-wk** [comment]
+10

**Pull request merged**